File: feature/src/main/java/com/ortin/gesturetranslator/feature/repository/LoadImageRepositoryImpl.java
package com.ortin.gesturetranslator.feature.repository;

import androidx.lifecycle.Lifecycle;
import androidx.lifecycle.LifecycleOwner;

import com.ortin.gesturetranslator.domain.listeners.LoadImagesListener;
import com.ortin.gesturetranslator.domain.models.Image;
import com.ortin.gesturetranslator.domain.repository.LoadImageRepository;
import com.ortin.gesturetranslator.feature.managers.camera.CameraManager;
import com.ortin.gesturetranslator.feature.managers.camera.listeners.CameraListener;
import com.ortin.gesturetranslator.feature.managers.camera.models.ImageFromCamera;

public class LoadImageRepositoryImpl implements LoadImageRepository {
    private final CameraManager cameraManager;

    public LoadImageRepositoryImpl(CameraManager cameraManager) {
        this.cameraManager = cameraManager;
    }

    @Override
    public void loadImages(LoadImagesListener loadImagesListener, LifecycleOwner lifecycleOwner) {
        cameraManager.loadImage(mapperToDomainListener(loadImagesListener), lifecycleOwner);
    }





    // Правила перевода для связи domain и feature модулей

    private Image mapToDomain(ImageFromCamera imageFromCamera) {
        return new Image(imageFromCamera.getImage(), imageFromCamera.getRotaion());
    }

    private CameraListener mapperToDomainListener(LoadImagesListener loadImagesListener) {
        return new CameraListener() {
            @Override
            public void getImage(ImageFromCamera imageFromCamera) {
                loadImagesListener.getImage(mapToDomain(imageFromCamera));
            }

            @Override
            public void error(Exception exception) {
                loadImagesListener.error(exception);
            }
        };
    }
}

File: feature/src/main/kotlin/com/ortin/gesturetranslator/feature/repository/LoadImageRepositoryImpl.kt
package com.ortin.gesturetranslator.feature.repository

import androidx.lifecycle.LifecycleOwner
import com.ortin.gesturetranslator.domain.listeners.LoadImagesListener
import com.ortin.gesturetranslator.domain.models.CameraFacingSettings
import com.ortin.gesturetranslator.domain.models.Image
import com.ortin.gesturetranslator.domain.repository.LoadImageRepository
import com.ortin.gesturetranslator.feature.managers.camera.CameraManager
import com.ortin.gesturetranslator.feature.managers.camera.listeners.CameraListener
import com.ortin.gesturetranslator.feature.managers.camera.models.ImageFromCamera
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class LoadImageRepositoryImpl @Inject constructor(private val cameraManager: CameraManager) :
    LoadImageRepository {
    override fun loadImages(
        loadImagesListener: LoadImagesListener,
        lifecycleOwner: LifecycleOwner,
        cameraFacing: CameraFacingSettings
    ) {
        cameraManager.loadImage(
            cameraListener = loadImagesListener.mapToDomainListener(),
            lifecycleOwner = lifecycleOwner,
            cameraFacing = cameraFacing.mode
        )
    }

    override fun setStatusFlashlight(mode: Boolean) {
        cameraManager.setStatusFlashlight(mode)
    }

    // Translation rules for domain and feature modules
    private fun ImageFromCamera.mapToDomain(): Image = Image(this.image, this.rotaion)

    private fun LoadImagesListener.mapToDomainListener(): CameraListener =
        object : CameraListener {
            override fun getImage(imageFromCamera: ImageFromCamera) {
                this@mapToDomainListener.getImage(imageFromCamera.mapToDomain())
            }

            override fun error(exception: Exception) {
                this@mapToDomainListener.error(exception)
            }
        }
}
